lib/nosedrum/interactor.ex

```diff
@@ -0,0 +1,143 @@
+
+  **Note** that Discord only supports nesting 3 levels deep, like `command -> subcommand group -> subcommand`.
+
+  ## Example path:
+  ```elixir
+  %{
+    {"castle", MyApp.CastleCommand.description()} =>
+      %{
+        {"prepare", "Prepare the castle for an attack."} => [],
+        {"open", "Open up the castle for traders and visitors."} => [],
+        # ...
+      }
+  }
+  ```
+  """
+  @type application_command_path ::
+          %{
+            {group_name :: String.t(), group_desc :: String.t()} => [
+              application_command_path | [Nosedrum.ApplicationCommand.option()]
+            ]
+          }
+
+  @typedoc """
+  The name or pid of the Interactor process.
+  """
+  @type name_or_pid :: atom() | pid()
+
+  @doc """
+  Handle an Application Command invocation.
+
+  This callback should be invoked upon receiving an interaction via the `:INTERACTION_CREATE` event.
+
+  ## Example using `Nosedrum.Interactor.Dispatcher`:
+  ```elixir
+  # In your `Nostrum.Consumer` file:
+  def handle_event({:INTERACTION_CREATE, interaction, _ws_state}) do
+    IO.puts "Got interaction"
+    Nosedrum.Interactor.Dispatcher.handle_interaction(interaction)
+  end
+  ```
+  """
+  @callback handle_interaction(interaction :: Interaction.t(), name_or_pid) :: :ok
+
+  @doc """
+  Add a new command under the given name or application command path. Returns `:ok` if successful, and
+  `{:error, reason}` otherwise.
+
+  If the command already exists, it will be overwritten.
+  """
+  @callback add_command(
+              name_or_path :: String.t() | application_command_path,
+              command_module :: module,
+              scope :: command_scope,
+              name_or_pid
+            ) ::
+              :ok | {:error, reason :: String.t()}
+
+  @doc """
+  Remove the command under the given name or application command path. Returns `:ok` if successful, and
+  `{:error, reason}` otherwise.
+
+  If the command does not exist, no error should be returned.
+  """
+  @callback remove_command(
+              name_or_path :: String.t() | application_command_path,
+              command_id :: Nostrum.Snowflake.t(),
+              scope :: command_scope,
+              name_or_pid
+            ) ::
+              :ok | {:error, reason :: String.t()}
+
+  @doc """
+  Responds to an Interaction with the values in the given `t:Nosedrum.ApplicationCommand.response()`. Returns `{:ok}` if
+  successful, and a `Nostrum.Api.error()` otherwise.
+  """
+  @spec respond(Interaction.t(), Nosedrum.ApplicationCommand.response()) ::
+          {:ok} | Nostrum.Api.error()
+  def respond(interaction, command_response) do
+    type =
+      command_response
+      |> Keyword.get(:type, :channel_message_with_source)
+      |> convert_callback_type()
+
+    data =
+      command_response
+      |> Keyword.take([:content, :embeds, :components, :tts?, :allowed_mentions])
+      |> Map.new()
+      |> put_flags(command_response)
```

Comment:
I had a crash appear here while testing:

```elixir
18:13:34.301 [error] GenServer Nosedrum.Interactor.Dispatcher terminating
** (BadArityError) #Function<0.3359280/1 in Nosedrum.Interactor.put_flags/2> with arity 1 called with 2 arguments ({:ephemeral?, 64}, %{content: "hi"})
    (elixir 1.12.3) lib/enum.ex:2397: anonymous fn/4 in Enum.reduce/3
    (stdlib 3.16.1) maps.erl:410: :maps.fold_1/3
    (nosedrum 0.3.0) lib/nosedrum/interactor.ex:120: Nosedrum.Interactor.respond/2
    (nosedrum 0.3.0) lib/nosedrum/interactor/dispatcher.ex:145: Nosedrum.Interactor.Dispatcher.handle_cast/2
    (stdlib 3.16.1) gen_server.erl:695: :gen_server.try_dispatch/4
    (stdlib 3.16.1) gen_server.erl:771: :gen_server.handle_msg/6
    (stdlib 3.16.1) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
```

I wrote a command module like this:
```elixir
defmodule Bolt.Cogs.Test do
  @behaviour Nosedrum.ApplicationCommand

  def description do
    "Test application commands."
  end

  def command(interaction) do
    [content: "hi"]
  end

  def type do
    :slash
  end
end
```

The cog is added like this:
```elixir
Nosedrum.Interactor.Dispatcher.add_command("test", Bolt.Cogs.Test, :global)
```

Reply:
Whoops! I forgot how to `Enum.reduce` apparently :sweat_smile: I tested with your command and it should be fixed now!